### PR TITLE
Added expand and upgrade activities in persistent storage dashboard

### PR DESCRIPTION
**Pull request by afreen23**
*Upgrade Activity*
Screenshot from 2020-01-21 09-37-27

*Expand Activity*
Screenshot from 2020-01-21 09-38-29

### CODE REVIEW

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/cluster-expand-activity.tsx [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import ActivityItem from '@console/shared/src/components/dashboard/activity-card/ActivityItem';
+import { K8sResourceKind } from '@console/internal/module/k8s';
+
+export const ClusterExpandActivity: React.FC = () => (
```
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/cluster-expand-activity.tsx]: We can reuse this too, but may be in future we want to modify these, hence keeping separate

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx [hunk 2]
```diff
@@ -21,12 +21,35 @@ import {
...
+
+const storageClusterResource: FirehoseResource = {
+  isList: true,
+  kind: referenceForModel(OCSServiceModel),
+  namespaced: false,
+  prop: 'storage-cluster',
+};
+
+export const getOCSSubscription = (subscriptions: FirehoseResult): K8sResourceKind => {
+  const itemsData: K8sResourceKind[] = _.get(subscriptions, 'data');
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]: you can use TS 3.7 optional chaining - applies for all usages of `_.get`

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/cluster-expand-activity.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/cluster-expand-activity.tsx]: Sure, this actually is fetching a k8s resource, i will make changes accordingly

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx (hunk 2 above)
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]: I was waiting for this, now I ma looking for all instances , to nail down with the hammer :D

Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]: Screenshot from 2020-01-21 18-29-05

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx [hunk 3]
```diff
@@ -66,6 +101,15 @@ const OngoingActivity = withDashboardResources(
       DATA_RESILIENCY_QUERY[StorageDashboardQuery.RESILIENCY_PROGRESS],
       'loadError',
     ]);
+
+    const subscriptions = _.get(resources, 'subs') as FirehoseResult;
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]: you can use optional chaining here too 
`resources?.subs`

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx [hunk 4]
```diff
@@ -66,6 +101,15 @@ const OngoingActivity = withDashboardResources(
...
+    const subscriptionsLoaded = subscriptions?.loaded;
+    const ocsSubscription: K8sResourceKind = getOCSSubscription(subscriptions);
+
+    const storageClusters = _.get(resources, 'storage-cluster') as FirehoseResult;
+    const storageClustersLoaded = storageClusters?.loaded;
+    const ocsCluster: K8sResourceKind = _.get(storageClusters, ['data', '0']);
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]:
```suggestion
    const ocsCluster = storageClusters?.data?.[0];
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx [hunk 5]
```diff
@@ -21,12 +21,35 @@ import {
...
+const storageClusterResource: FirehoseResource = {
+  isList: true,
+  kind: referenceForModel(OCSServiceModel),
+  namespaced: false,
+  prop: 'storage-cluster',
+};
+
+export const getOCSSubscription = (subscriptions: FirehoseResult): K8sResourceKind => {
+  const itemsData: K8sResourceKind[] = subscriptions?.data;
+  return _.find(itemsData, (item) => _.get(item, 'spec.name') === OCS_OPERATOR);
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]:
```suggestion
  return _.find(itemsData, (item) => item?.spec?.name === OCS_OPERATOR);
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx [hunk 6]
```diff
@@ -66,6 +101,15 @@ const OngoingActivity = withDashboardResources(
       DATA_RESILIENCY_QUERY[StorageDashboardQuery.RESILIENCY_PROGRESS],
       'loadError',
     ]);
+
+    const subscriptions = _.get(resources, 'subs') as FirehoseResult;
+    const subscriptionsLoaded = subscriptions?.loaded;
+    const ocsSubscription: K8sResourceKind = getOCSSubscription(subscriptions);
+
+    const storageClusters = _.get(resources, 'storage-cluster') as FirehoseResult;
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]:
```suggestion
    const storageClusters: FirehoseResult = resources?.['storage-cluster'];
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx [hunk 7]
```diff
@@ -51,12 +74,24 @@ const RecentEvent = withDashboardResources(
...
 const OngoingActivity = withDashboardResources(
-  ({ watchPrometheus, stopWatchPrometheusQuery, prometheusResults }: DashboardItemProps) => {
+  ({
+    watchPrometheus,
+    stopWatchPrometheusQuery,
+    watchK8sResource,
+    stopWatchK8sResource,
+    resources,
+    prometheusResults,
+  }: DashboardItemProps) => {
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]: no need to specify `DashboardItemProps`
```suggestion
  }) => {
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx [hunk 8]
```diff
@@ -76,9 +120,25 @@ const OngoingActivity = withDashboardResources(
       });
     }
 
+    if (isOCSUpgradeActivity(ocsSubscription)) {
+      resourceActivities.push({
+        resource: ocsSubscription,
+        timestamp: _.get(ocsSubscription, 'status.lastUpdated'),
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]:
```suggestion
        timestamp: ocsSubscription?.status?.lastUpdated,
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx [hunk 9]
```diff
@@ -76,9 +120,25 @@ const OngoingActivity = withDashboardResources(
...
+        loader: () => import('./ocs-upgrade-activity').then((m) => m.OCSUpgradeActivity),
+      });
+    }
+
+    if (isClusterExpandActivity(ocsCluster)) {
+      resourceActivities.push({
+        resource: ocsCluster,
+        timestamp: null,
```
[the remark this answers is not in the document]
Comment (rawagner) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]: cant you use `status.lastUpdated` as timestamp ?

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/ocs-upgrade-activity.tsx [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import ActivityItem from '@console/shared/src/components/dashboard/activity-card/ActivityItem';
+import { SubscriptionState } from '@console/operator-lifecycle-manager';
+import { K8sResourceKind } from '@console/internal/module/k8s';
+
+export const isOCSUpgradeActivity = (subscription: K8sResourceKind): boolean => {
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/ocs-upgrade-activity.tsx]: can be one liner and should reuse `getSubscriptionStatus` from `operator-lifecycle` plugin
```suggestion
export const isOCSUpgradeActivity = (subscription: SubscriptionKind): boolean => getSubscriptionStatus(subscription) === SubscriptionState.SubscriptionStateUpgradePending;
```

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/cluster-expand-activity.tsx [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+  <ActivityItem>Expanding OCS Cluster</ActivityItem>
+);
+
+export const isClusterExpandActivity = (storageCluster: K8sResourceKind): boolean => {
```
Comment (rawagner) [inline frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/cluster-expand-activity.tsx]:
```suggestion
export const isClusterExpandActivity = (storageCluster: K8sResourceKind): boolean => storageCluster?.status?.phase === 'Expanding'
```

File: frontend/packages/console-app/src/plugin.tsx [hunk 12]
```diff
@@ -81,6 +83,24 @@ const plugin: Plugin<ConsumedExtensions> = [
...
+    type: 'Dashboards/Overview/Activity/Resource',
+    properties: {
+      k8sResource: {
+        isList: true,
+        kind: referenceForModel(OCSServiceModel),
+        namespaced: false,
+        prop: 'storage-cluster',
+      },
+      isActivity: isClusterExpandActivity,
+      getTimestamp: null,
```
[the remark this answers is not in the document]
Comment (afreen23) [reply frontend/packages/console-app/src/plugin.tsx]: alright

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]: There is not any lastUpdated field in this type of CRD. Only creationtimestamp is present.

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx [hunk 13]
```diff
@@ -76,9 +120,25 @@ const OngoingActivity = withDashboardResources(
       });
     }
 
+    if (isOCSUpgradeActivity(ocsSubscription)) {
+      resourceActivities.push({
+        resource: ocsSubscription,
+        timestamp: _.get(ocsSubscription, 'status.lastUpdated'),
+        loader: () => import('./ocs-upgrade-activity').then((m) => m.OCSUpgradeActivity),
```
[the remark this answers is not in the document]
Comment (afreen23) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/activity-card.tsx]: I was not aware of that. Thanks!

File: frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/ocs-upgrade-activity.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (andybraren) [reply frontend/packages/ceph-storage-plugin/src/components/dashboard-page/storage-dashboard/activity-card/ocs-upgrade-activity.tsx]: I agree, to align with the recent change in Cluster Overview the `Storage` status in the storage dashboard's status card should probably become "Upgrading" in the future (FYI @yuvalgalanti). For now though (if this is merging before Friday) including it as an activity seems okay.

**Pull request merged**